aws/resource_aws_batch_compute_environment.go

```diff
@@ -0,0 +1,496 @@
+	CREATING  = "CREATING"
+	DELETING  = "DELETING"
+	UPDATING  = "UPDATING"
+	DELETED   = "DELETED"
+	VALID     = "VALID"
+	FAILED    = "FAILED"
+)
+
+var reComputeEnvironmentName = regexp.MustCompile(`^[A-Za-z0-9_]*$`)
+var reArnOfIamRole = regexp.MustCompile(`^arn:aws:iam::[0-9]{12}:role/.*$`)
+var reArnOfIamInstanceProfile = regexp.MustCompile(`^arn:aws:iam::[0-9]{12}:instance-profile/.*$`)
+
+func resourceAwsBatchComputeEnvironment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsBatchComputeEnvironmentCreate,
+		Read:   resourceAwsBatchComputeEnvironmentRead,
+		Update: resourceAwsBatchComputeEnvironmentUpdate,
+		Delete: resourceAwsBatchComputeEnvironmentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"compute_environment_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: isCorrentComputeEnvironmentName,
+			},
+			"compute_resources": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MinItems: 0,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"bid_percentage": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							ForceNew: true,
+						},
+						"desired_vcpus": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"ec2_key_pair": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"image_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"instance_role": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							ValidateFunc: isArnOfIamInstanceProfile,
+						},
+						"instance_type": {
+							Type:     schema.TypeSet,
+							Required: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"max_vcpus": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"min_vcpus": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"security_group_ids": {
+							Type:     schema.TypeSet,
+							Required: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"spot_iam_fleet_role": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: isArnOfIamRole,
+						},
+						"subnets": {
+							Type:     schema.TypeSet,
+							Required: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"tags": tagsSchema(),
+						"type": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							ValidateFunc: validation.StringInSlice([]string{EC2, SPOT}, true),
+						},
+					},
+				},
+			},
+			"service_role": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: isArnOfIamRole,
```

Comment:
It would be better to use the validateArn function, which has more guards to validate ARNs. Thoughts?